src/envs/behavior.py

```diff
@@ -441,26 +443,38 @@ def make_behavior_option(name: str, types: Sequence[Type], params_space: Box,
     def _policy(state: State, memory: Dict, _objects: Sequence[Object],
                 _params: Array) -> Action:
         assert "has_terminated" in memory
-        assert ("controller" in memory and memory["controller"] is not None
-                )  # must call initiable() first, and it must return True
+        assert memory.get(
+            "policy_controller"
+        ) is not None  # must call initiable() first, and it must return True
         assert not memory["has_terminated"]
-        action_arr, memory["has_terminated"] = memory["controller"](state, env)
+        action_arr, memory["has_terminated"] = memory["policy_controller"](
+            state, env)
         return Action(action_arr)
 
     def _initiable(state: State, memory: Dict, objects: Sequence[Object],
                    params: Array) -> bool:
         igo = [object_to_ig_object(o) for o in objects]
         assert len(igo) == 1
-        if memory.get("controller") is None:
+        if memory.get("policy_controller") is None:
             # We want to reset the state of the environment to
             # the state in the init state so that our options can
             # run RRT/plan from here as intended!
             if state.simulator_state is not None:
                 env.task.reset_scene(state.simulator_state)
-            controller = controller_fn(env, igo[0], params, rng=rng)
-            memory["controller"] = controller
+            policy_controller = controller_fn(env,
+                                              igo[0],
+                                              params,
+                                              ret_option_model=False,
+                                              rng=rng)
+            memory["policy_controller"] = policy_controller
+            model_controller = controller_fn(env,
```

Comment:
if I understand correctly, this is going to run planning twice -- once in creating the policy controller, and once in creating the model controller. is that right? can we avoid that if so?

this is making me want more the refactor where we avoid the controller functions altogether, and just create options directly in `behavior_options.py`... but if there's an easy way to address this concern in the current paradigm that's fine too, and maybe I or someone else can try the refactor in a later PR.

Reply:
Yes, that is exactly right (and something I discovered only after running it lol).

I don't really know an easy way to address this in the current paradigm unfortunately... Would love to refactor to create options directly, though not sure if we should merge this and do that in a separate PR?

Reply:
I'll leave it up to you (or @ronuchit if he feels strongly). If there's any chance that this refactor could also help clear up the bugs you're seeing with the current code, that would be reason to prioritize the refactor. But it sounds like you've localized the issue to the grasp sampler, so it might not.

Reply:
can we just make each controller fn cache the results of the planning/RRT part? that way it can be re-used?